### PR TITLE
[4650][ADD] stock_valuation_fifo_lot

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
[4650](https://www.quartile.co/web#id=4650&menu_id=505&cids=3&action=1457&model=project.task&view_type=form)

### CODE REVIEW

File: stock_valuation_fifo_lot/views/stock_valuation_layer_views.xml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    </record>
+
+    <record id="stock_valuation_layer_tree" model="ir.ui.view">
+        <field name="name">stock.valuation.layer.tree</field>
+        <field name="model">stock.valuation.layer</field>
+        <field name="inherit_id" ref="stock_account.stock_valuation_layer_tree" />
+        <field name="arch" type="xml">
+            <field name="quantity" position="after">
+                <field name="remaining_qty" optional="show" />
+            </field>
```
Comment (yostashiro) [inline stock_valuation_fifo_lot/views/stock_valuation_layer_views.xml]: IMO, showing `remaining_qty` and `remaining_value` fields in the list view in this module is a bit out of context and it should be done in another module (e.g. stock_valuation_layer_show_balance with its own context, or stock_view_adjust).  Otherwise this version of the module will diverge from the OCA which makes it difficult to control manage the module.

File: stock_valuation_fifo_lot/models/product.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    def _sort_by_all_candidates(self, all_candidates, sort_by):
+        """Hook function for other sort by"""
+        return all_candidates
+
+    def _get_all_candidates(self, company, sort_by=None):
```
Comment (yostashiro) [inline stock_valuation_fifo_lot/models/product.py]: @AungKoKoLin1997 Please follow up on your own comment https://github.com/OCA/stock-logistics-workflow/pull/1527#pullrequestreview-2174235639, first by creating a new PR in the OCA repo perhaps (I think we've given it enough time).

File: stock_valuation_fifo_lot/models/product.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                ]
+            )
+        )
+        if sort_by == "lot_create_date":
+
+            def sorting_key(candidate):
+                if len(candidate.lot_ids) > 1:
+                    return min(candidate.lot_ids.mapped("create_date"))
+                elif candidate.lot_ids:
+                    return candidate.lot_ids[0].create_date
```
Comment (yostashiro) [inline stock_valuation_fifo_lot/models/product.py]:
```suggestion
                if candidate.lot_ids:
                    return min(candidate.lot_ids.mapped("create_date"))
```

File: stock_valuation_fifo_lot/models/product.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            all_candidates = all_candidates.sorted(key=sorting_key)
+        elif sort_by is not None:
+            all_candidates = self._sort_by_all_candidates(all_candidates, sort_by)
+        return all_candidates
+
+    def _run_fifo(self, quantity, company):
+        self.ensure_one()
+        move_id = self._context.get("used_in_move_id")
+        if self.tracking == "none" or not move_id:
+            vals = super()._run_fifo(quantity, company)
```
Comment (yostashiro) [inline stock_valuation_fifo_lot/models/product.py]:
```suggestion
            return super()._run_fifo(quantity, company)
```

and then remove the else ladder to save some space and improve the readability.

File: stock_valuation_fifo_lot/models/product.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+            # Calculate standard price (Sorted by lot created date)
+            all_candidates = self._get_all_candidates(
+                company, sort_by="lot_create_date"
+            )
+            new_standard_price = 0.0
+            if all_candidates:
+                new_standard_price = all_candidates[0].unit_cost
+            elif candidates:
+                new_standard_price = candidate.unit_cost
```
Comment (yostashiro) [inline stock_valuation_fifo_lot/models/product.py]: Should this have been the intention? Please double check.
```suggestion
                new_standard_price = candidates[0].unit_cost
```

**Review by yostashiro — CHANGES_REQUESTED** [review]
@AungKoKoLin1997 Please do a thorough review of the code, checking the consistency against the latest 16.0 standard methods especially for big ones such as `_run_fifo()`.

We should also add tests.  I feel it's too risky to go as it is.

File: stock_valuation_fifo_lot/models/stock_move.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                        ("quantity", ">", 0),
+                        ("value", ">", 0),
+                        ("company_id", "=", self.company_id.id),
+                    ],
+                    limit=1,
+                )
+            )
+            if candidates:
+                price_unit = candidates[0].unit_cost
+        return price_unit
```
Comment (yostashiro) [inline stock_valuation_fifo_lot/models/stock_move.py]: I don't feel comfortable with the current logic. I think this logic is primarily for increasing the quantity of the existing lot in an inventory adjustment, IIUC.

I think the correct approach may be just to follow the standard logic, or do the following (ideally based on the configuration):

- Get remaining_value / remaining_qty from the candidate if remaining_qty is positive
- Otherwise, follow the standard logic

@AungKoKoLin1997 What do you think?

File: stock_valuation_fifo_lot/models/stock_move.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+                if all_candidates:
+                    move.product_id.sudo().with_company(
+                        move.company_id.id
+                    ).with_context(
+                        disable_auto_svl=True
+                    ).standard_price = all_candidates[
+                        0
+                    ].unit_cost
+            layers |= layer
+        return layers
```
Comment (yostashiro) [inline stock_valuation_fifo_lot/models/stock_move.py]: @AungKoKoLin1997 I actually have no idea what this adjustment is for. Can you please explain what it is.

**Review by yostashiro — COMMENTED** [review]
I was also not sure if the reported issue in this PR https://github.com/OCA/stock-logistics-workflow/pull/1350#issuecomment-1840407086 has already been addressed.

File: stock_valuation_fifo_lot/models/stock_move.py (hunk 6 above)
Comment (AungKoKoLin1997) [reply stock_valuation_fifo_lot/models/stock_move.py]: I believe this is also for using lot price that is used when we create incoming or outgoing transfers.

File: stock_valuation_fifo_lot/models/stock_move.py (hunk 7 above)
Comment (AungKoKoLin1997) [reply stock_valuation_fifo_lot/models/stock_move.py]: This adjustment is for using first lot product price when we receive this product.
Eg,
- Receive product with lot number (11111) for price 100
- Receive product with lot number (22222) for price 200

Now, the standard price of product is 100.
Then,

- Deliver product with lot number(11111).

After that, the standard price of product is 200.
We create another receipt.

- Receive product with lot number (11111) for price 300

We want to assign this price as standard cost because this lot is created first. So, the standard price is 300 with current design. The odoo standard behavior will keep 200 as standard price.

Comment (yostashiro) [reply stock_valuation_fifo_lot/models/stock_move.py]: > - Receive product with lot number (11111) for price 300

Is the assumption here to change the price of the PO line for the first receipt of this lot from 100 (or 200) to 300, and then receive on the same stock move line as the first receipt by increasing the done quantity (sounds like a corner-case scenario to me)?

> We want to assign this price as standard cost because this lot is created first. So, the standard price is 300 with current design. The odoo standard behavior will keep 200 as standard price.

I fail to see the point of doing this manipulation. What business goal does it try to address?

**Review by yostashiro — COMMENTED** [review]
@AungKoKoLin1997 Can you please check my updates and add tests for more patterns including the corner-case ones.

- Receive serials 001, 002 and 003 for an incoming move. Deliver 002, return 002 and deliver 002 again (it should work now).
- Receive serials 001 at JPY100 and 002 at JPY200. Deliver 001 and 002. Return 002, and the returned cost should be JPY200.
- Change (increase/decrease) qty_done value of a done move line (incoming and outgoing).
- Test force_fifo_lot_id field.

**Comment by AungKoKoLin1997** [comment]
Need to add test case for checking `force_fifo_lot_id` field. I have no idea at the moment for this.

File: stock_valuation_fifo_lot/models/stock_move.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+                    ("company_id", "=", self.company_id.id),
+                ],
+                order="id desc",
+                limit=1,
+            )
+            if move_line:
+                if move_line.qty_consumed:
+                    return move_line.value_consumed / move_line.qty_consumed
+                else:
+                    return move_line.value_remaining / move_line.qty_remaining
```
Comment (yostashiro) [inline stock_valuation_fifo_lot/models/stock_move.py]:
```suggestion
                return move_line.value_remaining / move_line.qty_remaining
```

File: stock_valuation_fifo_lot/models/stock_move.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        """No PO (e.g. customer returns) and get the price unit from the last consumed
+        incoming move line for the lot.
+        """
+        self.ensure_one()
+        if not self.company_id.use_lot_cost_for_new_stock:
+            return super()._get_price_unit()
+        if hasattr(self, "purchase_line_id") and self.purchase_line_id:
+            return super()._get_price_unit()
+        if self.product_id.cost_method == "fifo" and len(self.lot_ids) == 1:
+            # Get the last consumed incoming move line.
```
Comment (yostashiro) [inline stock_valuation_fifo_lot/models/stock_move.py]:
```suggestion
            # Get the most recent incoming move line for the lot.
```

File: stock_valuation_fifo_lot/models/stock_move_line.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+                rec.qty_remaining = rec.product_uom_id._compute_quantity(
+                    sum(remaining_qty_layers.mapped("remaining_qty")),
+                    rec.product_id.uom_id,
+                )
+                rec.value_remaining = (
+                    sum(remaining_qty_layers.mapped("remaining_value"))
+                    * sum(remaining_qty_layers.mapped("remaining_qty"))
+                    / rec.qty_remaining
+                )
+                continue
```
Comment (yostashiro) [inline stock_valuation_fifo_lot/models/stock_move_line.py]: We should update the README as well to explain the design.

```suggestion
                # Specifically for the case where qty_done of the outgoing stock move line with done state is reduced,
                # which creates a positive stock valuation layer for the outgoing move.
                layers = rec.move_id.stock_valuation_layer_ids.filtered(lambda l: l.remaining_qty > 0)
                if not layers:
                    continue
                rec.qty_remaining = rec.product_id.uom_id._compute_quantity(
                    sum(layers.mapped("remaining_qty")), rec.product_uom_id
                )
                rec.value_remaining = (
                    sum(layers.mapped("remaining_value"))
                    * sum(layers.mapped("remaining_qty"))
                    / rec.qty_remaining
                )
                continue
```

**Review by yostashiro — CHANGES_REQUESTED** [review]
Thanks Lin! Haven't reviewed tests yet.

File: stock_valuation_fifo_lot/models/stock_move_line.py (hunk 10 above)
Comment (AungKoKoLin1997) [reply stock_valuation_fifo_lot/models/stock_move_line.py]: We still need this code for the case of there is no remaining_qty in SVL.
Eg, 
- Receive 10
- Deliver 10
- Reduce deliver qty_done to 8
At the moment, the remaining qty of outgoing move line is 2.
Then,
- Deliver 2 
In this case, we need to clear existing remaining_qty.
```
if not remaining_qty_layers:
   rec.qty_remaining = 0
   rec.value_remaining = 0
   continue
```

File: stock_valuation_fifo_lot/models/product.py [hunk 11]
```diff
@@ -85,7 +85,7 @@ def _run_fifo(self, quantity, company):
         move_lines = correction_move_line or fifo_move._get_out_move_lines()
         for ml in move_lines:
             fifo_lot = ml.force_fifo_lot_id or ml.lot_id
-            ml_qty = fifo_move.product_uom._compute_quantity(ml.qty_done, self.uom_id)
+            ml_qty = ml.product_uom_id._compute_quantity(ml.qty_done, self.uom_id)
```
Comment (AungKoKoLin1997) [inline stock_valuation_fifo_lot/models/product.py]: I did this to solve the issue of below case.
- Receive 1 dozen (12 units in SVL)
- Deliver 1 dozen
With previous code, it will just deduct 1 unit from remaining qty of receiving SVL.

**Comment by yostashiro** [comment]
@AungKoKoLin1997 Currently qty_remaining and value_remaining are not updated for incoming receipts where you manually assign the lot number, and I haven't figured out how to fix it. It seems the layers are not found when the following compute is triggered.  Would be great if you can fix this.

```.py
    @api.depends(
        "lot_id",
        "qty_base",
        "qty_consumed",
        "move_id.stock_valuation_layer_ids",
        "move_id.stock_valuation_layer_ids.remaining_value",
    )
    def _compute_remaining_value(self):
        for rec in self:
            if not rec.product_id._is_fifo() or not rec.lot_id:
                continue
            rec.qty_remaining = rec.qty_base - rec.qty_consumed
            layers = rec.move_id.stock_valuation_layer_ids.filtered(
                lambda x: x.lot_ids in [rec.lot_id]
            )
            remaining_qty = sum(layers.mapped("remaining_qty"))
            if not remaining_qty:
                rec.qty_remaining = 0
                rec.value_remaining = 0
                continue
            rec.value_remaining = (
                sum(layers.mapped("remaining_value"))
                * rec.qty_remaining
                / remaining_qty
            )
```

File: stock_valuation_fifo_lot/models/product.py (hunk 11 above)
Comment (yostashiro) [reply stock_valuation_fifo_lot/models/product.py]: Isn't this the issue you mentioned about the Odoo core? If that is the case, we shouldn't be handling it in this module.

Anyway, I've changed the strategy to use the UoM of the product (not the move UoM) for all the added quantity columns in stock.move.line (for ease of balance comparison between stock valuation layers and stock move lines).

Comment (AungKoKoLin1997) [reply stock_valuation_fifo_lot/models/product.py]: The issue I mentioned was not decreasing `qty_done` after product is delivered.
This is for just receive and deliver with same `uom` that is not default `uom` of product.

**Review by kanda999 — APPROVED** [review]
Code and Functional check

File: stock_valuation_fifo_lot/models/stock_move.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+            # To prevent unknown creation of negative inventory.
+            if (
+                product.cost_method == "fifo"
+                and product.tracking != "none"
+                and layer.remaining_qty < 0
+            ):
+                raise UserError(
+                    _("Negative inventory is not allowed for product %s.")
+                    % product.display_name
+                )
```
Comment (kanda999) [inline stock_valuation_fifo_lot/models/stock_move.py:58]: I don't think we need to depend on stock_no_negative because of this part.

**Comment by jcdrubay** [comment]
You might want to reference somewhere that in version 18 of Odoo this is becoming native (from my understanding):

https://github.com/odoo/odoo/commit/2d933b83613ad52d76ab457201adecac6fcf184b

**Comment by yostashiro** [comment]
@AungKoKoLin1997 Can you please reflect Kanda's last commit in the OCA PR and add a test as well?

**Comment by AungKoKoLin1997** [comment]
> Can you please reflect Kanda's last commit in the OCA PR and add a test as well?

@yostashiro Done!

**Review by yostashiro — APPROVED** [review]
Code review.
